src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -203,15 +203,48 @@ public void AddToPendingBatch(ProvenBlockHeader provenBlockHeader, HashHeightPai
             if (this.saveAsyncLoopException != null)
                 throw this.saveAsyncLoopException;
 
+            List<KeyValuePair<int, ProvenBlockHeader>> itemsToRemove = null;
+
             lock (this.lockObject)
             {
+                if (this.pendingTipHashHeight != null && provenBlockHeader.HashPrevBlock != this.pendingTipHashHeight.Hash)
+                {
+                    // The latest proven header to be added is the most recent from consensus
+                    // and is always assumed to be the consensus tip.
+                    // If a none consecutive item is added then there may have been a reorg in the chain
+                    // this can happen when the node is in the middle of rewinding its consensus
+
+                    itemsToRemove = new List<KeyValuePair<int, ProvenBlockHeader>>();
+
+                    // Walk back the batch and remove all the blocks that are on the fork.
+                    KeyValuePair<int, ProvenBlockHeader> lastItem = this.pendingBatch.Last();
+                    while (provenBlockHeader.HashPrevBlock != lastItem.Value.GetHash())
```

Comment:
Whole method can be much shorter and easier to read: 

```
public void AddToPendingBatch(ProvenBlockHeader provenBlockHeader, HashHeightPair newTip)
{
    this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(provenBlockHeader), provenBlockHeader, nameof(newTip), newTip);

    Guard.Assert(provenBlockHeader.GetHash() == newTip.Hash);

    // Stop the consensus loop.
    if (this.saveAsyncLoopException != null)
        throw this.saveAsyncLoopException;

    lock (this.lockObject)
    {
        // If an item is already there this means a reorg happened.
        // We always assume the latest header belongs to the longest chain so just overwrite the previous values.
        this.pendingBatch.AddOrReplace(newTip.Height, provenBlockHeader);
        this.pendingTipHashHeight = newTip;

        this.Cache.AddOrUpdate(newTip.Height, provenBlockHeader, provenBlockHeader.HeaderSize);

        for (int currentHeight = newTip.Height + 1;; currentHeight++)
        {
            bool removed = this.pendingBatch.Remove(currentHeight);

            if (!removed)
                break;

            this.Cache.Remove(currentHeight);
        }
    }
}
```

Reply:
I will put the `remove` from cache inside the lock.

I like how you solved it by walking forwards (not backwards) but not sure its much better because you will have to remove an empty items once

Also the cache update is inside the lock regardless of there being a fork

Reply:
no, this will not work if we have in PendingBatch
5,6,9',10'

Reply:
The solution is just to add a check when to break the `if`, comparing it with height of the last PendingBatch element

Reply:
```
int maxHeight = this.pendingBatch.Keys.Last();
for (int currentHeight = newTip.Height + 1; i<= maxHeight; currentHeight++)
{
   bool removed = this.pendingBatch.Remove(currentHeight);
   this.Cache.Remove(currentHeight);
}
```